moveit_core/collision_detection/src/collision_env.cpp

```diff
@@ -0,0 +1,298 @@
+  : robot_model_(model), world_(new World()), world_const_(world_)
+{
+  if (!validateScale(scale))
+    scale = 1.0;
+  if (!validatePadding(padding))
+    padding = 0.0;
+
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    link_padding_[link->getName()] = padding;
+    link_scale_[link->getName()] = scale;
+  }
+}
+
+CollisionEnv::CollisionEnv(const robot_model::RobotModelConstPtr& model, const WorldPtr& world, double padding, double scale)
+  : robot_model_(model), world_(world), world_const_(world_)
+{
+  if (!validateScale(scale))
+    scale = 1.0;
+  if (!validatePadding(padding))
+    padding = 0.0;
+
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    link_padding_[link->getName()] = padding;
+    link_scale_[link->getName()] = scale;
+  }
+}
+
+CollisionEnv::CollisionEnv(const CollisionEnv& other, const WorldPtr& world) : robot_model_(other.robot_model_), world_(world), world_const_(world)
+{
+  link_padding_ = other.link_padding_;
+  link_scale_ = other.link_scale_;
+}
+void CollisionEnv::setPadding(double padding)
+{
+  if (!validatePadding(padding))
+    return;
+  std::vector<std::string> u;
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    if (getLinkPadding(link->getName()) != padding)
+      u.push_back(link->getName());
+    link_padding_[link->getName()] = padding;
+  }
+  if (!u.empty())
+    updatedPaddingOrScaling(u);
+}
+
+void CollisionEnv::setScale(double scale)
+{
+  if (!validateScale(scale))
+    return;
+  std::vector<std::string> u;
+  const std::vector<const robot_model::LinkModel*>& links = robot_model_->getLinkModelsWithCollisionGeometry();
+  for (auto link : links)
+  {
+    if (getLinkScale(link->getName()) != scale)
+      u.push_back(link->getName());
+    link_scale_[link->getName()] = scale;
+  }
+  if (!u.empty())
+    updatedPaddingOrScaling(u);
+}
+
+void CollisionEnv::setLinkPadding(const std::string& link_name, double padding)
+{
+  bool update = getLinkPadding(link_name) != padding;
+  link_padding_[link_name] = padding;
+  if (update)
+  {
+    std::vector<std::string> u(1, link_name);
+    updatedPaddingOrScaling(u);
+  }
+}
+
+double CollisionEnv::getLinkPadding(const std::string& link_name) const
+{
+  auto it = link_padding_.find(link_name);
+  if (it != link_padding_.end())
+    return it->second;
+  else
+    return 0.0;
+}
+
+void CollisionEnv::setLinkPadding(const std::map<std::string, double>& padding)
+{
+  std::vector<std::string> u;
+  for (const auto& link_pad_pair : padding)
+  {
+    bool update = getLinkPadding(link_pad_pair.first) != link_pad_pair.second;
+    link_padding_[link_pad_pair.first] = link_pad_pair.second;
```

Comment:
Same as above: should add `validatePadding` here.

Reply:
added it.